Add warnings blog post (#23592)
kubernetes/website: content/en/blog/_posts/2020-09-03-warnings/index.md

The warning is sent in a response header, so it does not change the status code or response body in any way.
This allows the server to send warnings while remaining compatible with previous client versions.

Warnings are surfaced by `kubectl` v1.19+ in `stderr` output, and by the `k8s.io/client-go` client library v0.19.0+ in log output.
The `k8s.io/client-go` behavior can be overridden [per-process](https://godoc.org/k8s.io/client-go/rest#SetDefaultWarningHandler)
or [per-client](https://godoc.org/k8s.io/client-go/rest#Config).

# Deprecation Warnings

The first way we are using this new capability is to send warnings for use of deprecated APIs.

Kubernetes is a [big, fast-moving project](https://www.cncf.io/cncf-kubernetes-project-journey/#development-velocity).
Keeping up with the [changes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.19.md#changelog-since-v1180)
in each release can be daunting, even for people who work on the project full-time. One important type of change is API deprecations.
As APIs in Kubernetes graduate to GA versions, pre-release API versions are deprecated and eventually removed.

> Even though there is an [extended deprecation period](/docs/reference/using-api/deprecation-policy/),

Reviewer: This deprecation policy url is 404ing for me. Sends me to `https://github.com/kubernetes/website/blob/1ff24eb921850f62180334cd210ec7d3bc3e9ab0/docs/reference/using-api/deprecation-policy`?

Author: links from the github preview aren't accurate

from https://deploy-preview-23592--kubernetes-io-master-staging.netlify.app/blog/2020/09/03/warnings/ it links to https://deploy-preview-23592--kubernetes-io-master-staging.netlify.app/docs/reference/using-api/deprecation-policy/ and works

Reviewer: Thank you, finally figured it out!

> Even though there is an [extended deprecation period](https://kubernetes.io/docs/reference/using-api/deprecation-policy/),

Reviewer 2: better only /docs/reference/using-api/deprecation-policy/

This warning includes details about the release in which the API will no longer be available, and the replacement API version.

Because the warning originates at the server, and is intercepted at the client level, it works for all kubectl commands,
including high-level commands like `kubectl apply`, and low-level commands like `kubectl get --raw`:

![kubectl applying a manifest file, then displaying a warning message 'networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress'.](kubectl-warnings.png)

This helps people affected by the deprecation to know the request they are making is deprecated,
how long they have to address the issue, and what API they should use instead.
This is especially helpful when the user is applying a manifest they didn't create,
so they have time to reach out to the authors to ask for an updated version.

We also realized that the person *using* a deprecated API is often not the same person responsible for upgrading the cluster,
so we added two administrator-facing tools to help track use of deprecated APIs and determine when upgrades are safe.

> 

Reviewer 3: I would expect to see the header name here for non-client-go clients.

Author: called that out and linked to the RFC in the intro

## Metrics 

Starting in Kubernetes v1.19, when a request is made to a deprecated REST API endpoint,
an `apiserver_requested_deprecated_apis` gauge metric is set to `1` in the kube-apiserver process.
This metric has labels for the API `group`, `version`, `resource`, and `subresource`,
and a `removed_version` label that indicates the Kubernetes release in which the API will no longer be served.

This is an example query using `kubectl`, [prom2json](https://github.com/prometheus/prom2json),

> and [jq](https://stedolan.github.io/jq/) to determine which deprecated APIs are being requested:

Reviewer 3: mention that this metrics info is temporary for that API server binary.

Author: done

  },
  "value": "1"
}
```

The output shows that only read requests are being made to these APIs, and the most requests have been made to watch the deprecated Ingress API.

This information could also be obtained via a Prometheus query.
The following Prometheus query returns information about requests made to deprecated APIs which will be removed in v1.22:

```promql
apiserver_requested_deprecated_apis{removed_version="1.22"} * on(group,version,resource,subresource)
group_right() apiserver_request_total
```

## Audit annotations

Metrics are a fast way to check whether deprecated APIs are being used, and at what rate,
but they don't include enough information to identify particular clients or API objects.
Starting in Kubernetes v1.19, [audit events](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/)
for requests to deprecated APIs include an audit annotation of `"k8s.io/deprecated":"true"`.
Administrators can use those audit events to identify specific clients or objects that need to be updated.

# Custom Resource Definitions

Along with the API server ability to warn about deprecated API use, starting in v1.19, a CustomResourceDefinition can indicate a 

> [particular version of the resource it defines is deprecated](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-deprecation).

Reviewer 2: same here can do it direct as /docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-deprecation

The author of the CustomResourceDefinition can also customize the warning for each version if they want to.
This allows them to give a pointer to a migration guide or other information if needed.

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
  name: crontabs.example.com
spec:
  versions:
  - name: v1alpha1
    # This indicates the v1alpha1 version of the custom resource is deprecated.
    # API requests to this version receive a warning in the server response.
    deprecated: true
    # This overrides the default warning returned to clients making v1alpha1 API requests.
    deprecationWarning: "example.com/v1alpha1 CronTab is deprecated; use example.com/v1 CronTab (see http://example.com/v1alpha1-v1)"
    ...

  - name: v1beta1
    # This indicates the v1beta1 version of the custom resource is deprecated.
    # API requests to this version receive a warning in the server response.
    # A default warning message is returned for this version.
    deprecated: true
    ...

  - name: v1
    ...
```

# Admission Webhooks

> [Admission webhooks](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers)

Reviewer 2: and another one /docs/reference/access-authn-authz/extensible-admission-controllers

---
layout: blog

> title: "Warning: Helpful Warnings Ahead"

Reviewer 4: No review, just ❤️

date: 2020-09-03
slug: warnings
---

**Author**: Jordan Liggitt (Google)

As Kubernetes maintainers, we're always looking for ways to improve usability while preserving compatibility.
As we develop features, triage bugs, and answer support questions, we accumulate information that would be helpful for Kubernetes users to know.
In the past, sharing that information was limited to out-of-band methods like release notes, announcement emails, documentation, and blog posts.
Unless someone knew to seek out that information and managed to find it, they would not benefit from it.

In Kubernetes v1.19, we added a feature that allows the Kubernetes API server to
[send warnings to API clients](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1693-warnings).

> The warning is sent using a [standard `Warning` response header](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.46),

Reviewer 5: ```suggestion
The warning is sent using a [standard `Warning` response header](https://tools.ietf.org/html/rfc7234#section-5.5),
```

(RFC 2616 has been superseded)

Author: updated

As APIs in Kubernetes graduate to GA versions, pre-release API versions are deprecated and eventually removed.

Even though there is an [extended deprecation period](/docs/reference/using-api/deprecation-policy/),
and deprecations are [included in release notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.19.md#deprecation),
they can still be hard to track. During the deprecation period, the pre-release API remains functional,
allowing several releases to transition to the stable API version. However, we have found that users often don't even realize
they are depending on a deprecated API version until they upgrade to the release that stops serving it.

Starting in v1.19, whenever a request is made to a deprecated REST API, a warning is returned along with the API response.
This warning includes details about the release in which the API will no longer be available, and the replacement API version.

Because the warning originates at the server, and is intercepted at the client level, it works for all kubectl commands,
including high-level commands like `kubectl apply`, and low-level commands like `kubectl get --raw`:

> ![kubectl applying a manifest file, then displaying a warning message 'networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress'.](kubectl-warnings.png)

Reviewer 5: nit: on a high-DPI screen, this screenshot looks blurry.

You could also use a terminal with fewer rows to save a bit of page real estate.

Don't let this feedback block a merge!

Author: weird... I intentionally took the screenshots on a high-DPI screen... not sure how to improve that

I'll crop the empty lines at the bottom

Reviewer 5: The blog stylesheet expands images to be the width of the page, so maybe this is a styling thing. If we never rendered the terminal image more than 80em wide for example, the pixels from the source image wouldn't render so large.

Something for another day, for sure.

Author: I think it's some netlify or css scaling... when I look at the images locally, they look sharp

'
```

Output:

```json
{
  "labels": {
    "code": "0",
    "component": "apiserver",
    "contentType": "application/vnd.kubernetes.protobuf;stream=watch",
    "dry_run": "",
    "group": "extensions",
    "resource": "ingresses",
    "scope": "cluster",
    "subresource": "",
    "verb": "WATCH",
    "version": "v1beta1"
  },
  "value": "21"
}
{
  "labels": {
    "code": "200",
    "component": "apiserver",
    "contentType": "application/vnd.kubernetes.protobuf",
    "dry_run": "",
    "group": "extensions",
    "resource": "ingresses",
    "scope": "cluster",
    "subresource": "",
    "verb": "LIST",
    "version": "v1beta1"
  },
  "value": "1"
}
{
  "labels": {
    "code": "200",
    "component": "apiserver",
    "contentType": "application/json",
    "dry_run": "",
    "group": "rbac.authorization.k8s.io",
    "resource": "clusterroles",
    "scope": "cluster",
    "subresource": "",
    "verb": "LIST",
    "version": "v1beta1"
  },
  "value": "1"
}
```

The output shows that only read requests are being made to these APIs, and the most requests have been made to watch the deprecated Ingress API.

> This information could also be obtained via a Prometheus query.

Reviewer 5: nit; how about:
```suggestion
You can also find that information through a Prometheus query,
```

Author: done

    ]
  }
}
```

If you are implementing a webhook that returns a warning message, here are some tips:

* Don't include a "Warning:" prefix in the message (that is added by clients on output)
* Use warning messages to describe problems the client making the API request should correct or be aware of
* Be brief; limit warnings to 120 characters if possible

There are many ways admission webhooks could use this new feature, and I'm looking forward to seeing what people come up with.
Here are a couple ideas to get you started:

* webhook implementations adding a "complain" mode, where they return warnings instead of rejections,
  to allow trying out a policy to verify it is working as expected before starting to enforce it
* "lint" or "vet"-style webhooks, inspecting objects and surfacing warnings when best practices are not followed

## Kubectl strict mode

If you want to be sure you notice deprecations as soon as possible and get a jump start on addressing them,
`kubectl` added a `--warnings-as-errors` option in v1.19. When invoked with this option,
`kubectl` treats any warnings it receives from the server as errors and exits with a non-zero exit code:

> ![kubectl applying a manifest file with a --warnings-as-errors flag, displaying a warning message and exiting with a non-zero exit code.](kubectl-warnings-as-errors.png)

Reviewer 5: nit: on a high-DPI screen, this screenshot looks blurry.

Love the alt attribute!

This could be used in a CI job to apply manifests to a current server,
and required to pass with a zero exit code in order for the CI job to succeed.

## Future Possibilities

Now that we have a way to communicate helpful information to users in context,
we're already considering other ways we can use this to improve people's experience with Kubernetes.
A couple areas we're looking at next are warning about [known problematic values](http://issue.k8s.io/64841#issuecomment-395141013)
we cannot reject outright for compatibility reasons, and warning about use of deprecated fields or field values

> (like selectors using beta os/arch node labels, [deprecated in v1.14](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.14.md#deprecations)).

Reviewer 5: You can if you want to link specifically to https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#beta-kubernetes-io-os-deprecated and https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#beta-kubernetes-io-arch-deprecated

Author: updated
